File: stepik-union/src/main/java/org/stepik/core/StudyUtils.java
package org.stepik.core;

import com.intellij.ide.ui.LafManager;
import com.intellij.ide.ui.laf.darcula.DarculaLookAndFeelInfo;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowManager;
import com.intellij.ui.content.Content;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.stepik.api.client.StepikApiClient;
import org.stepik.api.exceptions.StepikClientException;
import org.stepik.api.objects.StudyObject;
import org.stepik.api.objects.recommendations.Recommendation;
import org.stepik.api.objects.recommendations.Recommendations;
import org.stepik.api.objects.steps.Step;
import org.stepik.api.objects.steps.Steps;
import org.stepik.core.courseFormat.StudyNode;
import org.stepik.core.courseFormat.stepHelpers.StepHelper;
import org.stepik.core.templates.Templater;
import org.stepik.core.ui.StudyToolWindow;
import org.stepik.core.ui.StudyToolWindowFactory;
import org.stepik.core.utils.ProjectFilesUtils;

import javax.swing.*;
import java.util.HashMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import static org.stepik.core.stepik.StepikAuthManager.authAndGetStepikApiClient;
import static org.stepik.core.stepik.StepikAuthManager.isAuthenticated;

public class StudyUtils {
    private static final Logger logger = Logger.getInstance(StudyUtils.class);
    private static final String PATH_PATTERN = "^(?:(?:section([0-9]+)/lesson([0-9]+)/step([0-9]+))|" +
            "(?:lesson([0-9]+)/step([0-9]+))|" +
            "(?:step([0-9]+))|" +
            "(?:section([0-9]+)/lesson([0-9]+))|" +
            "(?:section([0-9]+))" +
            ").*$";
    private static Pattern pathPattern;

    private StudyUtils() {
    }

    static void initToolWindows(@NotNull final Project project) {
        final ToolWindowManager windowManager = ToolWindowManager.getInstance(project);
        windowManager.getToolWindow(StudyToolWindowFactory.STUDY_TOOL_WINDOW)
                .getContentManager()
                .removeAllContents(false);
        StudyToolWindowFactory factory = new StudyToolWindowFactory();
        factory.createToolWindowContent(project, windowManager.getToolWindow(StudyToolWindowFactory.STUDY_TOOL_WINDOW));
    }

    @Nullable
    static StudyToolWindow getStudyToolWindow(@NotNull final Project project) {
        if (project.isDisposed()) {
            return null;
        }
        ToolWindow toolWindow = ToolWindowManager.getInstance(project)
                .getToolWindow(StudyToolWindowFactory.STUDY_TOOL_WINDOW);
        if (toolWindow != null) {
            Content[] contents = toolWindow.getContentManager().getContents();
            for (Content content : contents) {
                JComponent component = content.getComponent();
                if (component != null && component instanceof StudyToolWindow) {
                    return (StudyToolWindow) component;
                }
            }
        }
        return null;
    }

    @NotNull
    public static String getStepContent(@NotNull StepHelper stepHelper) {
        return processTemplate(stepHelper, "quiz/" + stepHelper.getType());
    }

    @NotNull
    private static String processTemplate(@NotNull StepHelper stepHelper, @NotNull String templateName) {
        HashMap<String, Object> params = new HashMap<>();
        params.put("stepNode", stepHelper);
        params.put("darcula", LafManager.getInstance().getCurrentLookAndFeel() instanceof DarculaLookAndFeelInfo);

        return Templater.processTemplate(templateName, params);
    }

    @Nullable
    public static StudyPluginConfigurator getConfigurator(@NotNull final Project project) {
        StudyPluginConfigurator[] extensions = StudyPluginConfigurator.EP_NAME.getExtensions();
        for (StudyPluginConfigurator extension : extensions) {
            if (extension.accept(project)) {
                return extension;
            }
        }
        return null;
    }

    @NotNull
    private static String getRelativePath(@NotNull Project project, @NotNull VirtualFile item) {
        String path = item.getPath();
        String basePath = project.getBasePath();

        if (basePath == null) {
            return path;
        }

        return ProjectFilesUtils.getRelativePath(basePath, path);
    }

    @Nullable
    public static StudyNode getStudyNode(@NotNull Project project, @NotNull VirtualFile nodeVF) {
        String path = getRelativePath(project, nodeVF);

        StudyNode root = StepikProjectManager.getProjectRoot(project);
        if (root == null) {
            return null;
        }

        return getStudyNode(root, path);
    }

    @Nullable
    public static StudyNode getStudyNode(@NotNull StudyNode root, @NotNull String relativePath) {
        if (relativePath.equals(".")) {
            return root;
        }

        if (pathPattern == null) {
            pathPattern = Pattern.compile(PATH_PATTERN);
        }
        Matcher matcher = pathPattern.matcher(relativePath);
        if (!matcher.matches()) {
            return null;
        }

        for (int i = 1; i <= matcher.groupCount(); i++) {
            String idString = matcher.group(i);

            if (idString == null) {
                continue;
            }

            int id = Integer.parseInt(idString);

            root = root.getChildById(id);
            if (root == null) {
                return null;
            }
        }

        return root;
    }

    @Nullable
    public static StudyNode<?, ?> getRecommendation(@NotNull StudyNode root) {
        StudyObject data = root.getData();
        if (data == null || !data.isAdaptive()) {
            return null;
        }

        StudyNode studyNode = null;
        try {
            StepikApiClient stepikClient = authAndGetStepikApiClient();
            if (!isAuthenticated()) {
                return null;
            }
            Recommendations recommendations = stepikClient.recommendations()
                    .get()
                    .course(root.getId())
                    .execute();
            if (!recommendations.isEmpty()) {
                Recommendation recommendation = recommendations.getFirst();

                long lesson = recommendation.getLesson();

                Steps steps = stepikClient.steps()
                        .get()
                        .lesson(lesson)
                        .execute();
                if (!steps.isEmpty()) {
                    long stepId = steps.getFirst().getId();
                    studyNode = root.getChildByClassAndId(Step.class, stepId);
                }
            }
        } catch (StepikClientException e) {
            logger.warn(e);
        }
        return studyNode;
    }
}

File: stepik-core/src/main/kotlin/org/stepik/core/StudyUtils.kt
package org.stepik.core

import com.intellij.openapi.components.ServiceManager.getService
import com.intellij.openapi.project.Project
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.openapi.wm.ToolWindowManager
import org.stepik.api.exceptions.StepikClientException
import org.stepik.core.auth.StepikAuthManager.authAndGetStepikApiClient
import org.stepik.core.auth.StepikAuthManager.isAuthenticated
import org.stepik.core.common.Loggable
import org.stepik.core.courseFormat.StepNode
import org.stepik.core.courseFormat.StudyNode
import org.stepik.core.ui.StudyToolWindow
import org.stepik.core.ui.StudyToolWindowFactory
import org.stepik.core.ui.StudyToolWindowFactory.Companion.STUDY_TOOL_WINDOW
import org.stepik.core.utils.getRelativePath

private val logging = object : Loggable {}

private val pathPattern by lazy {
    ("^(?:(?:section([0-9]+)/lesson([0-9]+)/step([0-9]+))|" +
     "(?:lesson([0-9]+)/step([0-9]+))|" +
     "(?:step([0-9]+))|" +
     "(?:section([0-9]+)/lesson([0-9]+))|" +
     "(?:section([0-9]+))" +
     ").*$"
    ).toRegex()
}

fun initToolWindows(project: Project) {
    val toolWindow = ToolWindowManager.getInstance(project)
            .getToolWindow(STUDY_TOOL_WINDOW)
    toolWindow.contentManager.removeAllContents(false)
    StudyToolWindowFactory().createToolWindowContent(project, toolWindow)
}

fun getStudyToolWindow(project: Project): StudyToolWindow? {
    if (project.isDisposed) {
        return null
    }
    val toolWindowManager = ToolWindowManager.getInstance(project) ?: return null
    val contents = toolWindowManager.getToolWindow(STUDY_TOOL_WINDOW)
            ?.contentManager?.contents
    return contents?.mapNotNull { it.component as? StudyToolWindow }
            ?.firstOrNull()
}

fun getConfigurator(project: Project): StudyPluginConfigurator? {
    return getProjectManager(project)?.getConfigurator(project)
}

private fun getRelativePath(project: Project, item: VirtualFile): String {
    val basePath = project.basePath ?: return item.path
    return basePath.getRelativePath(item.path)
}

fun getStudyNode(project: Project, nodeVF: VirtualFile): StudyNode? {
    val root = getProjectRoot(project) ?: return null
    val path = getRelativePath(project, nodeVF)
    
    return getStudyNode(root, path)
}

fun getStudyNode(root: StudyNode, relativePath: String): StudyNode? {
    if (relativePath == ".") {
        return root
    }
    
    val matcher = pathPattern.matchEntire(relativePath) ?: return null
    
    return matcher.groupValues.drop(1)
            .filterNot { it.isEmpty() }
            .map { it.toLong() }
            .fold(root as StudyNode?) { newRoot, id ->
                newRoot?.getChildById(id)
            }
}

fun getProjectManager(project: Project?): ProjectManager? {
    project ?: return null
    return getService(project, ProjectManager::class.java)
}

fun getProjectRoot(project: Project?): StudyNode? {
    return getProjectManager(project)?.projectRoot
}

fun isStepikProject(project: Project?): Boolean {
    return getProjectRoot(project) != null
}

fun getRecommendation(root: StudyNode): StudyNode? {
    val data = root.data
    if (!data.isAdaptive) {
        return null
    }
    
    var studyNode: StudyNode? = null
    
    val stepikClient = authAndGetStepikApiClient()
    if (!isAuthenticated) {
        return null
    }
    
    try {
        val recommendations = stepikClient.recommendations()
                .get()
                .course(root.id)
                .execute()
        if (recommendations.isNotEmpty) {
            val lesson = recommendations.first()
                    .lesson
            
            val step = stepikClient.steps()
                    .get()
                    .lesson(lesson)
                    .execute()
                    .firstOrNull()
            if (step != null) {
                val stepId = step.id
                studyNode = root.getChildByClassAndId(StepNode::class.java, stepId)
            }
        }
    } catch (e: StepikClientException) {
        logging.logger.warn(e)
    }
    
    return studyNode
}

val pluginId = getService(PluginSettings::class.java).pluginId

val pluginMetricsName = getService(PluginSettings::class.java).pluginMetricsName

val host = getService(PluginSettings::class.java).host

val clientId = getService(PluginSettings::class.java).clientId

val loadCurrentUser = getService(PluginSettings::class.java)::currentUser

val pluginName = getService(PluginSettings::class.java).pluginName
